More cleanup of new framework defaults

prathamesh-sonpatki opened the pull request:
- Move real new default options to the top of the file.
- After that club together all the options which were added to keep
  backward compatibility. So all of them will get only one header.
- Based on https://github.com/rails/rails/pull/25231#issuecomment-222945173.

r? @kaspth

prathamesh-sonpatki committed:
More cleanup of new framework defaults

- Move real new default options to the top of the file.
- After that club together all the options which were added to keep
  backward compatibility. So all of them will get only one header.
- Based on https://github.com/rails/rails/pull/25231#issuecomment-222945173.

prathamesh-sonpatki commented:
Now we get output like this - 

``` sh
$ edgerails new devrails --dev --force --skip-active-record --api

$ cat devrails/config/initializers/new_framework_defaults.rb
# Be sure to restart your server when you modify this file.
# This file contains all the new default configuration options from
# Rails 5.0.

# Following config options are introduced to ensure that apps
# made on earlier versions of Rails are not affected when upgrading.

# Do not halt callback chains when a callback returns false.
ActiveSupport.halt_callback_chains_on_return_false = false

# Configure SSL options to enable HSTS with subdomains.
Rails.application.config.ssl_options = { hsts: { subdomains: true } }

# Preserve the timezone of the receiver when calling to `to_time`.
# Ruby 2.4 will change the behavior of `to_time` to preserve the timezone
# when converting to an instance of `Time` instead of the previous behavior
# of converting to the local system timezone.
ActiveSupport.to_time_preserves_timezone = true
```

``` sh
$ edgerails new devrails --dev --force --skip-active-record
$ cat devrails/config/initializers/new_framework_defaults.rb
# Be sure to restart your server when you modify this file.
# This file contains all the new default configuration options from
# Rails 5.0.

# Enable per-form CSRF tokens.
Rails.application.config.action_controller.per_form_csrf_tokens = true

# Enable origin-checking CSRF mitigation.
Rails.application.config.action_controller.forgery_protection_origin_check = true

# Following config options are introduced to ensure that apps
# made on earlier versions of Rails are not affected when upgrading.

# Do not halt callback chains when a callback returns false.
ActiveSupport.halt_callback_chains_on_return_false = false

# Configure SSL options to enable HSTS with subdomains.
Rails.application.config.ssl_options = { hsts: { subdomains: true } }

# Preserve the timezone of the receiver when calling to `to_time`.
# Ruby 2.4 will change the behavior of `to_time` to preserve the timezone
# when converting to an instance of `Time` instead of the previous behavior
# of converting to the local system timezone.
ActiveSupport.to_time_preserves_timezone = true
```

```sh
$ edgerails new devrails --dev --force --api
$ cat devrails/config/initializers/new_framework_defaults.rb
# Be sure to restart your server when you modify this file.
# This file contains all the new default configuration options from
# Rails 5.0.

# Following config options are introduced to ensure that apps
# made on earlier versions of Rails are not affected when upgrading.

# Require `belongs_to` associations by default.
Rails.application.config.active_record.belongs_to_required_by_default = true

# Do not halt callback chains when a callback returns false.
ActiveSupport.halt_callback_chains_on_return_false = false

# Configure SSL options to enable HSTS with subdomains.
Rails.application.config.ssl_options = { hsts: { subdomains: true } }

# Preserve the timezone of the receiver when calling to `to_time`.
# Ruby 2.4 will change the behavior of `to_time` to preserve the timezone
# when converting to an instance of `Time` instead of the previous behavior
# of converting to the local system timezone.
ActiveSupport.to_time_preserves_timezone = true
```

```sh
$ edgerails new devrails --dev --force
$ cat devrails/config/initializers/new_framework_defaults.rb 
# Be sure to restart your server when you modify this file.
# This file contains all the new default configuration options from
# Rails 5.0.

# Enable per-form CSRF tokens.
Rails.application.config.action_controller.per_form_csrf_tokens = true

# Enable origin-checking CSRF mitigation.
Rails.application.config.action_controller.forgery_protection_origin_check = true

# Following config options are introduced to ensure that apps
# made on earlier versions of Rails are not affected when upgrading.

# Require `belongs_to` associations by default.
Rails.application.config.active_record.belongs_to_required_by_default = true

# Do not halt callback chains when a callback returns false.
ActiveSupport.halt_callback_chains_on_return_false = false

# Configure SSL options to enable HSTS with subdomains.
Rails.application.config.ssl_options = { hsts: { subdomains: true } }

# Preserve the timezone of the receiver when calling to `to_time`.
# Ruby 2.4 will change the behavior of `to_time` to preserve the timezone
# when converting to an instance of `Time` instead of the previous behavior
# of converting to the local system timezone.
ActiveSupport.to_time_preserves_timezone = true
```

prathamesh-sonpatki commented:
Build failures are unrelated.

kaspth commented:
Here's my take on what the file should say:

```ruby
# Be sure to restart your server when you modify this file.
#
# This file contains migration options to ease your Rails 5.0 upgrade.
# They will be removed in the next Rails version.
#
# Flip the defaults to the value of past versions then once upgraded
# begin migrating to the new default one by one.
#
# Read the Rails 5.0 release notes for more info on each option.

# Enable per-form CSRF tokens. Previous versions had false.
Rails.application.config.action_controller.per_form_csrf_tokens = true

# Enable origin-checking CSRF mitigation. Previous versions had false.
Rails.application.config.action_controller.forgery_protection_origin_check = true

# Require `belongs_to` associations by default. Previous versions had false.
Rails.application.config.active_record.belongs_to_required_by_default = true

# Halt callback chains when a callback returns `false`. Previous versions had true.
ActiveSupport.halt_callback_chains_on_return_false = false

# Configure SSL options to enable HSTS with subdomains. Previous versions had false.
Rails.application.config.ssl_options = { hsts: { subdomains: true } }

# Make Ruby 2.4 preserve the timezone of the receiver when calling `to_time`.
# Previous versions had false.
ActiveSupport.to_time_preserves_timezone = true
```

I think it's a good idea to print out past version defaults so people have an easier time seeing how far they are in their migration.

However I wonder how much we should put in the top of the file intro. @dhh do you have any opinions on this? Overthinking or a worthy UX improvement?

dhh commented:
"They will be removed in the next Rails version" should probably be "They will be removed in the next major Rails version".

Otherwise good 👍

prathamesh-sonpatki commented:
Great. I will update it soon :)

kaspth commented:
Sweet! Yeah, let's add "major" in there. I didn't use it at first because I thought people might conflate it with Rails 6.0 when they could be removed in Rails 5.1.

prathamesh-sonpatki committed:
Fix minor regression about old apps not getting per_form_csrf and request_forgery_protection configs

- Earlier per_form_csrf_tokens and request_forgery_protection config
  files were generated for old apps upgraded to Rails 5.
- But when we collapsed all initializers into one file, the entire file
  does not get created for old apps.
- This commit fixes it and also changes values for all new defaults for
  old apps so that they will not break.
- Also added a test for `rails app:update`.

prathamesh-sonpatki committed:
Update the documentation of new_framework_defaults file with more details about old apps and how you can upgrade to new defaults.

[Kasper Timm Hansen, Prathamesh Sonpatki]

prathamesh-sonpatki commented:
@kaspth Regarding 

> Flip the defaults to the value of past versions

I added https://github.com/rails/rails/pull/25235/commits/e72b2d0a859883ce754c12d3bd58135ff74a5c47 to generate proper values for older apps getting upgraded to Rails 5. So we will *by default* generate flipped values for older apps. Let me know your thoughts on it.

kaspth commented on railties/lib/rails/tasks/framework.rake:
```diff
@@ -48,7 +48,7 @@ namespace :app do
           require 'rails/generators/rails/app/app_generator'
           gen = Rails::Generators::AppGenerator.new ["rails"],
-                                                    { api: !!Rails.application.config.api_only },
+                                                    { api: !!Rails.application.config.api_only, update: true, force: ENV['FORCE'] },
```
Was force just added for tests? What does it mean for end users?

prathamesh-sonpatki replied:
Yes. This is only for tests. The other option was manually create the generator class and pass `update: true` but I chose to call `bin/rails app:update` directly from tests.

In the long term, I would definitely like to move the update task to new command line infrastructure :)

kaspth replied:
I really don't like the idea of adding this option since end users won't need it.

The tests already created the AppGenerator manually, what's the harm in continuing that? 😁

prathamesh-sonpatki replied:
Sure. I updated it to 
``` ruby
generator = Rails::Generators::AppGenerator.new ["rails"], { update: true }, destination_root: app_root, shell: @shell
```

Notice the `update: true` part.

kaspth commented on railties/lib/rails/generators/rails/app/templates/config/initializers/new_framework_defaults.rb.tt:
```diff
@@ -1,37 +1,36 @@
 Rails.application.config.action_controller.per_form_csrf_tokens = true
 
-# Enable origin-checking CSRF mitigation.
+# Enable origin-checking CSRF mitigation. This is new feature of Rails 5.
```
True these are new features but enabling them when upgrading has ramifications on apps. Let's wrap in update and set to false on upgrading apps; new apps get these features enabled.

Then for consistency, let's add `Previous versions had false`.

prathamesh-sonpatki replied:
@kaspth Not sure if this will cause issues with older apps. I haven't dug into the implications. I will check once more closely.

prathamesh-sonpatki replied:
Checked the documentation 

> * `config.action_controller.forgery_protection_origin_check` configures whether the HTTP `Origin` header should be checked against the site's origin as an additional CSRF defense.

I think it is safe to not enable for older apps. I will update as per your suggestion.

kaspth commented:
@prathamesh-sonpatki I ❤️❤️❤️ the upgrade switch. Making sure upgrading apps keep the previous default should quench a lot of upgrading woes.

kaspth commented on railties/lib/rails/generators/rails/app/app_generator.rb:
```diff
@@ -102,10 +101,6 @@ def config_when_updating
 
-      unless new_framework_defaults_config_exist
-        remove_file 'config/initializers/new_framework_defaults.rb'
-      end
```
Wait did this code say remove the file unless it exists? Well... 🤓

kaspth commented on railties/lib/rails/generators/rails/app/templates/config/initializers/new_framework_defaults.rb.tt:
```diff
@@ -1,37 +1,36 @@
 ActiveSupport.to_time_preserves_timezone = true
+
+# Following config options are introduced to ensure that apps
+# made on earlier versions of Rails are not affected when upgrading.
```
Think this isn't true as the defaults introduced with new features in Rails 5 have ramifications on upgrading apps. Let's just remove this because the whole file has become about ensuring this 😁

prathamesh-sonpatki replied:
Yeah. No longer needed now.

kaspth commented on railties/lib/rails/generators/rails/app/templates/config/initializers/new_framework_defaults.rb.tt:
```diff
@@ -1,37 +1,36 @@
+
+# Do not halt callback chains when a callback returns false. Previous versions had true.
+ActiveSupport.halt_callback_chains_on_return_false = <%= options[:update] ? true : false %>
+<%- unless options[:update] -%>
```
Defaulting to subdomains true will still be a default in 5.1 right? Then I think we should mirror the other configs; output this but flip the value.

prathamesh-sonpatki replied:
Yes. In 5.1 we will default to `subdomains: true`. The problem with this is older apps might have `Rails.application.config.ssl_options` set to something else. That was the reason for not adding this change for older apps. 

They will get a deprecation warning and hopefully set the config to some value before 5.1 --> https://github.com/rails/rails/blob/b24d44edbb4726d06df782dc5b4b0e52db467d97/actionpack/lib/action_dispatch/middleware/ssl.rb#L-69-L-73

kaspth commented on railties/lib/rails/generators/rails/app/templates/config/initializers/new_framework_defaults.rb.tt:
```diff
@@ -1,37 +1,36 @@
+# This file contains migration options to ease your Rails 5.0 upgrade.
+# They will be removed in the next major Rails version.
+#
+# Flip the defaults to the value of past versions then once upgraded
```
Now handled automatically. Let's say: `Once upgraded flip defaults one by one to migrate to the new default.`

prathamesh-sonpatki replied:
Sure.

kaspth commented on railties/lib/rails/generators/rails/app/templates/config/initializers/new_framework_defaults.rb.tt:
```diff
@@ -1,37 +1,36 @@
+# made on earlier versions of Rails are not affected when upgrading.
+<%- unless options[:skip_active_record] -%>
+
+# Require `belongs_to` associations by default. Previous versions had false.
```
Since we helpfully output the past version for upgraders, this is less helpful. Let's say `Next major release defaults to true.` here and in the others. That ought to give people a sense a sense of where they ought to go next, and people can rest easy seeing the already have the right default.

prathamesh-sonpatki replied:
@kaspth We can generate this only for older apps now. New apps can have `Previous versions had false.`.

kaspth replied:
> We can generate this only for older apps now

I'm not sure I understand?

prathamesh-sonpatki committed:
Remove FORCE option as it was only needed for tests

- Adjusted tests instead.

prathamesh-sonpatki committed:
Use the `update` flag to provide more friendlier messages to people upgrading.

prathamesh-sonpatki commented:
Updated as per @kaspth's suggestions.

Older app upgraded to Rails 5 get this now - 

```
$ cat config/initializers/new_framework_defaults.rb 
# Be sure to restart your server when you modify this file.
#
# This file contains migration options to ease your Rails 5.0 upgrade.
# They will be removed in the next major Rails version.
#
# Once upgraded flip defaults one by one to migrate to the new default.
#
# Read the Rails 5.0 release notes for more info on each option.

# Enable per-form CSRF tokens. Next major version defaults to true.
Rails.application.config.action_controller.per_form_csrf_tokens = false

# Enable origin-checking CSRF mitigation. Next major version defaults to true.
Rails.application.config.action_controller.forgery_protection_origin_check = false

# Make Ruby 2.4 preserve the timezone of the receiver when calling `to_time`.
ActiveSupport.to_time_preserves_timezone = true

# Require `belongs_to` associations by default. Next major version defaults to true.
Rails.application.config.active_record.belongs_to_required_by_default = false

# Do not halt callback chains when a callback returns false. Next major version defaults to false.
ActiveSupport.halt_callback_chains_on_return_false = true
```

New apps - 

```
$ edgerails new devrails --dev --force
$ cat devrails/config/initializers/new_framework_defaults.rb 
# Be sure to restart your server when you modify this file.
#
# This file contains migration options to ease your Rails 5.0 upgrade.
# They will be removed in the next major Rails version.
#
# Read the Rails 5.0 release notes for more info on each option.

# Enable per-form CSRF tokens. Previous versions had false.
Rails.application.config.action_controller.per_form_csrf_tokens = true

# Enable origin-checking CSRF mitigation. Previous versions had false.
Rails.application.config.action_controller.forgery_protection_origin_check = true

# Make Ruby 2.4 preserve the timezone of the receiver when calling `to_time`.
ActiveSupport.to_time_preserves_timezone = true


# Require `belongs_to` associations by default. Previous versions had false.
Rails.application.config.active_record.belongs_to_required_by_default = true

# Do not halt callback chains when a callback returns false. Previous versions had true.
ActiveSupport.halt_callback_chains_on_return_false = false

# Configure SSL options to enable HSTS with subdomains. Previous versions had false.
Rails.application.config.ssl_options = { hsts: { subdomains: true } }
```

```
$ edgerails new devrails --dev --force --skip-active-record
$ cat devrails/config/initializers/new_framework_defaults.rb 
# Be sure to restart your server when you modify this file.
#
# This file contains migration options to ease your Rails 5.0 upgrade.
# They will be removed in the next major Rails version.
#
# Read the Rails 5.0 release notes for more info on each option.

# Enable per-form CSRF tokens. Previous versions had false.
Rails.application.config.action_controller.per_form_csrf_tokens = true

# Enable origin-checking CSRF mitigation. Previous versions had false.
Rails.application.config.action_controller.forgery_protection_origin_check = true

# Make Ruby 2.4 preserve the timezone of the receiver when calling `to_time`.
ActiveSupport.to_time_preserves_timezone = true

# Do not halt callback chains when a callback returns false. Previous versions had true.
ActiveSupport.halt_callback_chains_on_return_false = false

# Configure SSL options to enable HSTS with subdomains. Previous versions had false.
Rails.application.config.ssl_options = { hsts: { subdomains: true } }
```

```
$ edgerails new devrails --dev --api --force

$ cat devrails/config/initializers/new_framework_defaults.rb 
# Be sure to restart your server when you modify this file.
#
# This file contains migration options to ease your Rails 5.0 upgrade.
# They will be removed in the next major Rails version.
#
# Read the Rails 5.0 release notes for more info on each option.

# Make Ruby 2.4 preserve the timezone of the receiver when calling `to_time`.
ActiveSupport.to_time_preserves_timezone = true

# Require `belongs_to` associations by default. Previous versions had false.
Rails.application.config.active_record.belongs_to_required_by_default = true

# Do not halt callback chains when a callback returns false. Previous versions had true.
ActiveSupport.halt_callback_chains_on_return_false = false

# Configure SSL options to enable HSTS with subdomains. Previous versions had false.
Rails.application.config.ssl_options = { hsts: { subdomains: true } }
```

```
$ edgerails new devrails --dev --api --skip-active-record --force
$ cat devrails/config/initializers/new_framework_defaults.rb 
# Be sure to restart your server when you modify this file.
#
# This file contains migration options to ease your Rails 5.0 upgrade.
# They will be removed in the next major Rails version.
#
# Read the Rails 5.0 release notes for more info on each option.

# Make Ruby 2.4 preserve the timezone of the receiver when calling `to_time`.
ActiveSupport.to_time_preserves_timezone = true

# Do not halt callback chains when a callback returns false. Previous versions had true.
ActiveSupport.halt_callback_chains_on_return_false = false

# Configure SSL options to enable HSTS with subdomains. Previous versions had false.
Rails.application.config.ssl_options = { hsts: { subdomains: true } }
```

I tested the config for CodeTriage.com locally which is upgraded from Rails 4 to Rails 5. The flipped defaults seem to working fine. Though I had to do lot of changs related to gems to accommodate Rails 5.1.0.alpha. But that is not related to this PR.

@kaspth Please review.

sgrif commented on railties/lib/rails/generators/rails/app/templates/config/initializers/new_framework_defaults.rb.tt:
```diff
@@ -1,37 +1,34 @@
-Rails.application.config.active_record.belongs_to_required_by_default = true
+#
+# This file contains migration options to ease your Rails 5.0 upgrade.
+# They will be removed in the next major Rails version.
```
This isn't necessarily true. We don't need to mention this here, each of the setters will have a deprecation warning individually before they are removed.

prathamesh-sonpatki replied:
@sgrif :+1: Removed it.

kaspth commented:
Backported to 5-0-stable @ 5687ee0
